File: bubbler_app/lib/features/graph/widgets/neighbor_bubble.dart
import 'package:flutter/material.dart';

import '../../../data/models/graph.dart';
import '../../../data/models/topics.dart';
import '../../../shared/theme/topic_style.dart';

/// A single neighbor choice bubble — Swift `GraphNeighborBubble`.
class NeighborBubble extends StatelessWidget {
  const NeighborBubble({
    super.key,
    required this.node,
    required this.size,
    required this.onTap,
  });

  final GraphFeedNode node;
  final double size;
  final VoidCallback onTap;

  String get _topic => node.topicName ?? 'topic';

  @override
  Widget build(BuildContext context) {
    final color = TopicStyle.color(_topic);
    final displayName = KnownTopics.displayName(_topic);

    return Semantics(
      button: true,
      label:
          '$displayName bubble${node.isPreferredTopic ? ', preferred topic' : ''}',
      onTap: onTap,
      child: ExcludeSemantics(
        child: Material(
          type: MaterialType.transparency,
          child: InkWell(
            customBorder: const CircleBorder(),
            onTap: onTap,
            child: SizedBox(
              width: size,
              height: size,
              child: Stack(
                alignment: Alignment.center,
                children: [
                  DecoratedBox(
                    decoration: BoxDecoration(
                      shape: BoxShape.circle,
                      gradient: RadialGradient(
                        center: const Alignment(-0.6, -0.6),
                        radius: 0.95,
                        colors: [
                          color.withValues(alpha: 0.95),
                          color.withValues(alpha: 0.55),
                          color.withValues(alpha: 0.25),
                        ],
                        stops: const [0.0, 0.45, 1.0],
                      ),
                      border: Border.all(
                        color: Colors.white.withValues(alpha: 0.55),
                        width: 1.5,
                      ),
                      boxShadow: [
                        BoxShadow(
                          color: color.withValues(alpha: 0.45),
                          blurRadius: 14,
                          offset: const Offset(0, 8),
                        ),
                      ],
                    ),
                    child: const SizedBox.expand(),
                  ),
                  Padding(
                    padding: const EdgeInsets.symmetric(horizontal: 8),
                    child: Column(
                      mainAxisAlignment: MainAxisAlignment.center,
                      children: [
                        Icon(
                          TopicStyle.icon(_topic),
                          size: size * 0.22,
                          color: Colors.white,
                        ),
                        SizedBox(height: size * 0.06),
                        Text(
                          displayName,
                          maxLines: 1,
                          overflow: TextOverflow.ellipsis,
                          textAlign: TextAlign.center,
                          style: TextStyle(
                            fontSize: (size * 0.11).clamp(10.0, 18.0),
                            fontWeight: FontWeight.w700,
                            color: Colors.white,
                            height: 1.1,
                          ),
                        ),
                      ],
                    ),
                  ),
                  if (node.isPreferredTopic)
                    Positioned(
                      top: size * 0.08,
                      right: size * 0.08,
                      child: Container(
                        padding: EdgeInsets.all(size * 0.07),
                        decoration: BoxDecoration(
                          shape: BoxShape.circle,
                          color: Colors.white.withValues(alpha: 0.92),
                        ),
                        child: Icon(
                          Icons.star,
                          size: (size * 0.13).clamp(10.0, 20.0),
                          color: Colors.yellow.withValues(alpha: 0.9),
                        ),
                      ),
                    ),
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }
}

File: BubblerAndroid/app/src/main/java/com/bubbler/android/features/graph/components/NeighborBubble.kt
package com.bubbler.android.features.graph.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Star
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.bubbler.android.app.theme.BubblerTheme
import com.bubbler.android.data.model.GraphFeedNode
import com.bubbler.android.data.model.KnownTopics
import com.bubbler.android.data.model.Post
import com.bubbler.android.ui.theme.TopicStyle
import java.time.Instant

/**
 * Tappable neighbor choice bubble — mirrors Swift `GraphNeighborBubble`.
 */
@Composable
fun NeighborBubble(
    node: GraphFeedNode,
    size: Dp,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
) {
    val topic = node.topicName ?: "topic"
    val color = TopicStyle.color(topic)
    val displayName = KnownTopics.displayName(topic)
    val accessibilityLabel = buildString {
        append(displayName)
        append(" bubble")
        if (node.isPreferredTopic) append(", preferred topic")
    }

    val density = LocalDensity.current
    val sizePx = with(density) { size.toPx() }
    val iconSize = size * 0.22f
    val labelSp = maxOf(10f, sizePx * 0.11f / density.density)
    val starSize = maxOf(10.dp, size * 0.13f)
    val badgePadding = size * 0.07f
    val badgeInset = size * 0.08f

    Box(
        modifier = modifier
            .size(size)
            .semantics { contentDescription = accessibilityLabel }
            .shadow(
                elevation = 14.dp,
                shape = CircleShape,
                ambientColor = color.copy(alpha = 0.45f),
                spotColor = color.copy(alpha = 0.45f),
            )
            .clip(CircleShape)
            .background(
                Brush.radialGradient(
                    colors = listOf(
                        color.copy(alpha = 0.95f),
                        color.copy(alpha = 0.55f),
                        color.copy(alpha = 0.25f),
                    ),
                    center = Offset(
                        x = sizePx * 0.28f,
                        y = sizePx * 0.28f,
                    ),
                    radius = sizePx * 0.7f,
                ),
            )
            .border(1.5.dp, Color.White.copy(alpha = 0.55f), CircleShape)
            .clickable(enabled = enabled, onClick = onClick),
        contentAlignment = Alignment.Center,
    ) {
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(6.dp),
            modifier = Modifier.padding(horizontal = 8.dp),
        ) {
            Icon(
                imageVector = TopicStyle.icon(topic),
                contentDescription = null,
                tint = Color.White,
                modifier = Modifier.size(iconSize),
            )
            Text(
                text = displayName,
                color = Color.White,
                fontSize = labelSp.sp,
                fontWeight = FontWeight.Bold,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
        }

        if (node.isPreferredTopic) {
            Icon(
                imageVector = Icons.Filled.Star,
                contentDescription = null,
                tint = Color.Yellow.copy(alpha = 0.9f),
                modifier = Modifier
                    .align(Alignment.TopEnd)
                    .padding(badgeInset)
                    .size(starSize + badgePadding * 2)
                    .clip(CircleShape)
                    .background(Color.White.copy(alpha = 0.92f))
                    .padding(badgePadding),
            )
        }
    }
}

@Preview(showBackground = true, backgroundColor = 0xFF0D47A1)
@Composable
private fun NeighborBubblePreview() {
    BubblerTheme {
        NeighborBubble(
            node = GraphFeedNode(
                post = Post(
                    id = "bubble-1",
                    userId = 1,
                    username = "neighbor",
                    content = "Sample neighbor post",
                    createdAt = Instant.now(),
                    topic = "technology",
                ),
                isPreferredTopic = true,
            ),
            size = 96.dp,
            onClick = {},
            modifier = Modifier.padding(24.dp),
        )
    }
}

@Preview(showBackground = true, backgroundColor = 0xFF0D47A1)
@Composable
private fun NeighborBubblePlainPreview() {
    BubblerTheme {
        NeighborBubble(
            node = GraphFeedNode(
                post = Post(
                    id = "bubble-2",
                    userId = 2,
                    username = "walker",
                    content = "Another neighbor",
                    createdAt = Instant.now(),
                    topic = "sports",
                ),
            ),
            size = 96.dp,
            onClick = {},
            modifier = Modifier.padding(24.dp),
        )
    }
}
